util/topsql/stmtstats/stmtstats.go

```diff
@@ -0,0 +1,224 @@
+// to initialize the top-sql, nothing will happen, and we will get nil. But
+// this scene should never appear, because we always call SetupTopSQL before
+// starting the server, at this moment we cannot receive connections and will
+// not create a valid session. So this case will never happen: "This function
+// returns nil, so this valid session will count nothing".
+func CreateStatementStats() *StatementStats {
+	if manager == nil {
+		return nil
+	}
+	stats := &StatementStats{
+		data:   StatementStatsMap{},
+		closed: atomic.NewBool(false),
+	}
+	manager.register(stats)
+	return stats
+}
+
+// GetOrCreateStatementStatsItem creates the corresponding StatementStatsItem
+// for the specified SQLPlanDigest and timestamp if it does not exist before.
+// GetOrCreateStatementStatsItem is just a helper function, not responsible for
+// concurrency control, so GetOrCreateStatementStatsItem is **not** thread-safe.
+func (s *StatementStats) GetOrCreateStatementStatsItem(sqlDigest, planDigest string, ts int64) *StatementStatsItem {
+	key := SQLPlanDigest{SQLDigest: sqlDigest, PlanDigest: planDigest}
+	tsItem, ok := s.data[key]
+	if !ok {
+		s.data[key] = map[int64]*StatementStatsItem{}
+		tsItem = s.data[key]
+	}
+	item, ok := tsItem[ts]
+	if !ok {
+		tsItem[ts] = NewStatementStatsItem()
+		item = tsItem[ts]
+	}
+	return item
+}
+
+// AddExecCount is used to count the number of executions of a certain
+// SQLPlanDigest within a certain timestamp.
+// AddExecCount is thread-safe.
+func (s *StatementStats) AddExecCount(sqlDigest, planDigest string, ts int64, n uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	item := s.GetOrCreateStatementStatsItem(sqlDigest, planDigest, ts)
+	item.ExecCount += n
+}
+
+// AddKvExecCount is used to count the number of executions of a certain
+// SQLPlanDigest for a certain target within a certain timestamp。
+// AddKvExecCount is thread-safe.
+func (s *StatementStats) AddKvExecCount(sqlDigest, planDigest string, ts int64, target string, n uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	item := s.GetOrCreateStatementStatsItem(sqlDigest, planDigest, ts)
+	item.KvStatsItem.KvExecCount[target] += n
+}
+
+// Take removes all existing StatementStatsMap data from StatementStats.
+// Take is thread-safe.
+func (s *StatementStats) Take() StatementStatsMap {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	data := s.data
+	s.data = StatementStatsMap{}
+	return data
+}
+
+// Close marks StatementStats as "closed".
+// The background goroutine will periodically detect whether each
+// StatementStats has been closed, if so, it will be cleaned up.
+func (s *StatementStats) Close() {
+	s.closed.Store(true)
+}
+
+// Closed returns whether the StatementStats has been closed.
+func (s *StatementStats) Closed() bool {
+	return s.closed.Load()
+}
+
+// SQLPlanDigest is used as the key of StatementStatsMap to
+// distinguish different sql.
+type SQLPlanDigest struct {
+	SQLDigest  string
+	PlanDigest string
+}
```

Comment:
If I remember correctly @crazycs520 mainly use `[]byte` elsewhere. Maybe you need to unify them.

Personally I would prefer it to be `[]byte`, as there is possibility that people will feed `hex(digest)` into the string if it is `string`. While `[]byte` looks more like something binary.

Reply:
`[]byte` cannot be directly used as the key of map, because it does not implement `==` and `!=`, structs containing `[]byte` may have the same problem. cc @crazycs520

Reply:
Since we need to use `SQLPlanDigest` as the key of the map to merge `StatementStatsItem`, string must be used here. If we use []byte we will get the following compile error:
```go
type K struct {
	k1 []byte
	k2 []byte
}
m := map[K]int{}

// Invalid map key type: comparison operators == and != must be fully defined for the key type
```